src/shims/unix/socket.rs

```diff
@@ -22,15 +29,71 @@ impl FileDescription for SocketPair {
     ) -> InterpResult<'tcx, io::Result<()>> {
         Ok(Ok(()))
     }
+
+    fn read<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &mut [u8],
+        _tcx: TyCtxt<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let request_byte_size = bytes.len();
+        let data_size = self.readbuf.borrow_mut().len();
+
+        // FIXME: blocking is currently not supported
+        if data_size == 0 {
+            throw_unsup_format!("socketpair read: blocking isn't supported yet");
+        }
+        if request_byte_size == 0 {
+            return Ok(Ok(0));
+        }
+        // If the size requested > data_size in buffer, we only read up to data_size in buffer.
+        if request_byte_size > data_size {
+            for byte in bytes.iter_mut().take(data_size) {
+                *byte = self.readbuf.borrow_mut().pop_front().unwrap();
+            }
+            return Ok(Ok(data_size));
+        } else {
+            for byte in bytes.iter_mut() {
+                *byte = self.readbuf.borrow_mut().pop_front().unwrap();
+            }
+            return Ok(Ok(request_byte_size));
+        }
+    }
+
+    fn write<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &[u8],
+        _tcx: TyCtxt<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let available_space = self.max_capacity.checked_sub(self.writebuf.borrow().len()).unwrap();
+        let write_size = bytes.len();
+
+        // If there is no space left in the buffer, we will block until there is available space.
+        // FIXME: blocking is currently not supported.
+        if available_space == 0 {
+            throw_unsup_format!("socketpair write: blocking isn't supported yet");
+        }
+        if write_size == 0 {
+            return Ok(Ok(0));
+        }
```

Comment:
shouldn't these two checks be swapped? writing zero bytes into a full buffer is fine